src/main/fc/rc.c

```diff
@@ -263,23 +263,41 @@ static void scaleRawSetpointToFpvCamAngle(void)
 
 void updateRcRefreshRate(timeUs_t currentTimeUs)
 {
-    static timeUs_t lastRxTimeUs;
-
-    timeDelta_t frameAgeUs;
-    timeDelta_t frameDeltaUs = rxGetFrameDelta(&frameAgeUs);
+    // this function runs when a new frame is detected, after 150ms of no frames, and then every 50ms until the link is restored.
+    // it provides values for use in RCSmoothing, Feedforward, etc.
+
+    // get the frame interval reported by the Rx driver code
+    // this will be zero for NULL values and for forced checks in the absence of incoming data
+    timeDelta_t frameDeltaUsRx = rxGetFrameDelta();
+
+    // calculate frame interval here as well both on new packets and on forced checks without new packets
+    // provides an alternative delta value if zero returned by rxGetFrameDelta()
+    static timeUs_t lastRxTimeUs = 0;
+    timeDelta_t frameDeltaUs = cmpTimeUs(currentTimeUs, lastRxTimeUs);
+    if (rxIsReceivingSignal()) {
+        lastRxTimeUs = currentTimeUs;
```

Comment:
This function is either called when new frame is received (and currentTimeUs is used to time received frames) or at periodic intervals when no signal is received. But only rxIsReceivingSignal is used to distinguish two cases, making code flow very confusing.

Also, currentRxRateHz / currentRxIntervalUs seems to be used suspiciously in smoothing code ...

Reply:
`rxIsReceivingSignal` is true for up to 150ms after any packet.  While true, we should calculate delta and should update the previous frame time whenever a new frame triggers this function.

However after 150ms, `rxIsReceivingSignal` goes false, we enter Stage 1 failsafe, and, even though there are no new packets arriving, we re-check at fixed intervals, to monitor for stick movement during GPS rescue, for example, but also to update the interval since the last frame as the dropout gets longer. While the absence of new frames persists, we do not want to update the time stamp of the last good frame, otherwise the measured delta time won't reflect how long it has been since we received that last good frame.

Should I add some comments here?  
Should `lastRxTimeUs` be re-named to `previousFrameTimeUs`?

As to the 'suspicious' smoothing code, yes, indeed it is suspicious, mostly because it works well despite being suspicious.  `isRxIntervalValid` is the gate-keeper, disregarding any intervals outside the allowed min/max range.  Please advise how that could be improved.

Reply:
I agree that the whole Rx / failsafe code should be much less complex.

This is at least a small simplification, one step along that path.  And it is supported by data.

We could probably do most of these timing measurements inside of RxFrameCheck. 

The other parts 'work as they should' when tested carefully.  The problem with significant refactoring is how easy it could be to break it.

Reply:
It is quite OK, but I don't like rxIsReceivingSignal() hidden path. Just make it argument (see my PR), so at least the confusion is moved one level up


> The problem with significant refactoring is how easy it could be to break it.

There are unused code paths, redundant code, hidden data flow, masked problems, broken code. How long until this becomes totally unmaintainable?